File: wireframe/src/main/java/ru/nsu/e/shelbogashev/wireframe/bsplineeditor/SplinePanel.java
package ru.nsu.e.shelbogashev.wireframe.bsplineeditor;

import ru.nsu.e.shelbogashev.wireframe.utils.Point2D;
import ru.nsu.e.shelbogashev.wireframe.utils.Settings;

import javax.swing.*;
import java.awt.*;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.awt.event.MouseMotionListener;
import java.awt.event.MouseWheelEvent;
import java.awt.event.MouseWheelListener;

/**
 * Панель для редактирования B-сплайнов.
 */
public class SplinePanel extends JPanel implements MouseMotionListener, MouseListener, MouseWheelListener {
    private final BSpline bSpline;
    private final JScrollPane scrollPane = new JScrollPane();
    private int indent = Settings.DEFAULT_INDENT;
    private int curX;
    private int curY;
    private int activePointIndex;
    private Action curAction = Action.NOTHING;
    private int width = 640;
    private int height = 240;

    /**
     * Создает панель для редактирования B-сплайнов с заданными параметрами.
     *
     * @param spline Параметры B-сплайнов.
     */
    public SplinePanel(BSpline spline) {
        setPreferredSize(new Dimension(width, height));
        setBackground(Color.BLACK);
        bSpline = spline;
        scrollPane.setMaximumSize(new Dimension(width, height));
        scrollPane.setViewportView(this);
        scrollPane.revalidate();
        addMouseListener(this);
        addMouseMotionListener(this);
        addMouseWheelListener(this);
    }

    private void drawAxis(Graphics g) {
        g.setColor(Color.WHITE);
        g.drawLine(0, getHeight() / 2, getWidth(), getHeight() / 2);
        g.drawLine(getWidth() / 2, 0, getWidth() / 2, getHeight());
        int height = getHeight();
        int width = getWidth();
        int linesNumber = (width / 2) / indent;
        for (int i = 1; i <= linesNumber; i++) {
            g.drawLine(width / 2 - indent * i, height / 2 - 3, width / 2 - indent * i, height / 2 + 3);
            g.drawLine(width / 2 + indent * i, height / 2 - 3, width / 2 + indent * i, height / 2 + 3);
        }
        linesNumber = (height / 2) / indent;
        for (int i = 1; i <= linesNumber; i++) {
            g.drawLine(width / 2 - 3, height / 2 - indent * i, width / 2 + 3, height / 2 - indent * i);
            g.drawLine(width / 2 - 3, height / 2 + indent * i, width / 2 + 3, height / 2 + indent * i);
        }
    }

    private void drawSplineAnchorPoints(Graphics g) {
        g.setColor(Color.GREEN);
        var points = bSpline.getAnchorPoints();
        int centerX = getWidth() / 2;
        int centerY = getHeight() / 2;
        for (Point2D p : points) {
            g.drawOval((int) (centerX + p.getX() * indent - Settings.RADIUS / 2), (int) (centerY - p.getY() * indent - Settings.RADIUS / 2), Settings.RADIUS, Settings.RADIUS);
        }
        g.setColor(Color.WHITE);
        drawLinesBetweenPoints(g, points);
    }

    private void drawLinesBetweenPoints(Graphics g, java.util.List<Point2D> points) {
        for (int i = 0; i < points.size() - 1; i++) {
            Point2D p1 = convertToScreen(points.get(i));
            Point2D p2 = convertToScreen(points.get(i + 1));
            g.drawLine((int) p1.getX(), (int) p1.getY(), (int) p2.getX(), (int) p2.getY());
        }
    }

    private void drawSpline(Graphics g) {
        g.setColor(Color.RED);
        var points = bSpline.getSplinePoints();
        if (points == null || points.size() < 4)
            return;
        drawLinesBetweenPoints(g, points);
    }

    private Point2D convertToScreen(Point2D p) {
        int centerX = getWidth() / 2;
        int centerY = getHeight() / 2;
        double x = centerX + p.getX() * indent;
        double y = centerY - p.getY() * indent;
        return new Point2D(x, y);
    }

    /**
     * Пересоздает B-сплайн и перерисовывает панель.
     */
    public void recreateSpline() {
        bSpline.createBSpline();
        repaint();
    }

    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        drawAxis(g);
        drawSplineAnchorPoints(g);
        drawSpline(g);
    }

    @Override
    public void mouseDragged(MouseEvent e) {
        double centerX = (double) getWidth() / 2;
        double centerY = (double) getHeight() / 2;
        switch (curAction) {
            case POINT_MOVING -> {
                var curPoint = bSpline.getAnchorPoints().get(activePointIndex);
                curPoint.setX((e.getX() - centerX) / indent);
                curPoint.setY((centerY - e.getY()) / indent);
                bSpline.createBSpline();
            }
            case CANVAS_MOVING -> {
                int dx = curX - e.getX();
                int dy = curY - e.getY();
                Dimension panelSize = new Dimension(getWidth(), getHeight());
                Point viewPos = scrollPane.getViewport().getViewPosition();
                int maxX = viewPos.x + scrollPane.getViewport().getWidth();
                int maxY = viewPos.y + scrollPane.getViewport().getHeight();
                if (dx < 0 && viewPos.x == 0) {
                    width -= dx;
                    curX -= dx;
                } else if (dx >= 0 && maxX == panelSize.width) {
                    width += dx;
                }
                if (dy < 0 && viewPos.y == 0) {
                    height -= dy;
                    curY -= dy;
                } else if (dy >= 0 && maxY == panelSize.height) {
                    height += dy;
                }
                scrollPane.getHorizontalScrollBar().setValue(viewPos.x + dx);
                scrollPane.getVerticalScrollBar().setValue(viewPos.y + dy);
            }
        }
        repaint();
    }

    @Override
    public void mouseMoved(MouseEvent e) {
    }

    @Override
    public void mouseClicked(MouseEvent e) {
        double centerX = (double) getWidth() / 2;
        double centerY = (double) getHeight() / 2;
        if (e.getButton() == MouseEvent.BUTTON1)
            bSpline.addAnchorPoint(new Point2D((e.getX() - centerX) / indent, (centerY - e.getY()) / indent));
        if (e.getButton() == MouseEvent.BUTTON3)
            bSpline.deleteAnchorPoint();
        bSpline.createBSpline();
        repaint();
    }

    @Override
    public void mousePressed(MouseEvent e) {
        curX = e.getX();
        curY = e.getY();
        checkPoint();
    }

    private void checkPoint() {
        var anchorPoints = bSpline.getAnchorPoints();
        int centerX = getWidth() / 2;
        int centerY = getHeight() / 2;
        int i = 0;
        for (Point2D p : anchorPoints) {
            double globalX = p.getX() * indent + centerX;
            double globalY = centerY - p.getY() * indent;
            if (curX <= globalX + Settings.RADIUS && curX >= globalX - Settings.RADIUS && curY <= globalY + Settings.RADIUS && curY >= globalY - Settings.RADIUS) {
                activePointIndex = i;
                curAction = Action.POINT_MOVING;
                break;
            }
            i++;
        }
    }

    @Override
    public void mouseReleased(MouseEvent e) {
        curAction = Action.NOTHING;
        repaint();
    }

    @Override
    public void mouseEntered(MouseEvent e) {
    }

    @Override
    public void mouseExited(MouseEvent e) {
    }

    @Override
    public void mouseWheelMoved(MouseWheelEvent e) {
        int modifier = e.getWheelRotation();
        indent -= modifier;
        repaint();
    }

    enum Action {
        POINT_MOVING,
        CANVAS_MOVING,
        NOTHING
    }
}

File: wireframe/src/main/java/ru/nsu/e/shelbogashev/wireframe/bsplineeditor/SplinePanel.kt
package ru.nsu.e.shelbogashev.wireframe.bsplineeditor

import ru.nsu.e.shelbogashev.wireframe.utils.Point2D
import ru.nsu.e.shelbogashev.wireframe.utils.Settings
import java.awt.Color
import java.awt.Dimension
import java.awt.Graphics
import java.awt.event.*
import javax.swing.JPanel
import javax.swing.JScrollPane

/**
 * Панель для редактирования B-сплайнов.
 */
class SplinePanel(spline: BSpline) : JPanel(), MouseMotionListener, MouseListener, MouseWheelListener {
    private val bSpline: BSpline
    private val scrollPane = JScrollPane()
    private var indent = Settings.DEFAULT_INDENT
    private var curX = 0
    private var curY = 0
    private var activePointIndex = 0
    private var curAction = Action.NOTHING
    private var width = 640
    private var height = 240

    /**
     * Создает панель для редактирования B-сплайнов с заданными параметрами.
     */
    init {
        preferredSize = Dimension(width, height)
        background = Color.BLACK
        bSpline = spline
        scrollPane.maximumSize = Dimension(width, height)
        scrollPane.setViewportView(this)
        scrollPane.revalidate()
        addMouseListener(this)
        addMouseMotionListener(this)
        addMouseWheelListener(this)
    }

    private fun drawAxis(g: Graphics) {
        g.color = Color.WHITE
        g.drawLine(0, getHeight() / 2, getWidth(), getHeight() / 2)
        g.drawLine(getWidth() / 2, 0, getWidth() / 2, getHeight())
        val height = getHeight()
        val width = getWidth()
        var linesNumber = (width / 2) / indent
        for (i in 1..linesNumber) {
            g.drawLine(width / 2 - indent * i, height / 2 - 3, width / 2 - indent * i, height / 2 + 3)
            g.drawLine(width / 2 + indent * i, height / 2 - 3, width / 2 + indent * i, height / 2 + 3)
        }
        linesNumber = (height / 2) / indent
        for (i in 1..linesNumber) {
            g.drawLine(width / 2 - 3, height / 2 - indent * i, width / 2 + 3, height / 2 - indent * i)
            g.drawLine(width / 2 - 3, height / 2 + indent * i, width / 2 + 3, height / 2 + indent * i)
        }
    }

    private fun drawSplineAnchorPoints(g: Graphics) {
        g.color = Color.GREEN
        val points = bSpline.getAnchorPoints()
        val centerX = getWidth() / 2
        val centerY = getHeight() / 2
        for (p in points) {
            g.drawOval(
                (centerX + p.x * indent - Settings.RADIUS / 2).toInt(),
                (centerY - p.y * indent - Settings.RADIUS / 2).toInt(),
                Settings.RADIUS,
                Settings.RADIUS
            )
        }
        g.color = Color.WHITE
        drawLinesBetweenPoints(g, points)
    }

    private fun drawLinesBetweenPoints(g: Graphics, points: List<Point2D>) {
        for (i in 0 until points.size - 1) {
            val p1 = convertToScreen(points[i])
            val p2 = convertToScreen(points[i + 1])
            g.drawLine(p1.x.toInt(), p1.y.toInt(), p2.x.toInt(), p2.y.toInt())
        }
    }

    private fun drawSpline(g: Graphics) {
        g.color = Color.RED
        val points = bSpline.splinePoints
        if (points == null || points.size < 4) return
        drawLinesBetweenPoints(g, points)
    }

    private fun convertToScreen(p: Point2D): Point2D {
        val centerX = getWidth() / 2
        val centerY = getHeight() / 2
        val x = centerX + p.x * indent
        val y = centerY - p.y * indent
        return Point2D(x, y)
    }

    /**
     * Пересоздает B-сплайн и перерисовывает панель.
     */
    fun recreateSpline() {
        bSpline.createBSpline()
        repaint()
    }

    override fun paintComponent(g: Graphics) {
        super.paintComponent(g)
        drawAxis(g)
        drawSplineAnchorPoints(g)
        drawSpline(g)
    }

    override fun mouseDragged(e: MouseEvent) {
        val centerX = getWidth().toDouble() / 2
        val centerY = getHeight().toDouble() / 2
        when (curAction) {
            Action.POINT_MOVING -> {
                val curPoint = bSpline.getAnchorPoints()[activePointIndex]
                curPoint.x = (e.x - centerX) / indent
                curPoint.y = (centerY - e.y) / indent
                bSpline.createBSpline()
            }

            Action.CANVAS_MOVING -> {
                val dx = curX - e.x
                val dy = curY - e.y
                val panelSize = Dimension(getWidth(), getHeight())
                val viewPos = scrollPane.viewport.viewPosition
                val maxX = viewPos.x + scrollPane.viewport.width
                val maxY = viewPos.y + scrollPane.viewport.height
                if (dx < 0 && viewPos.x == 0) {
                    width -= dx
                    curX -= dx
                } else if (dx >= 0 && maxX == panelSize.width) {
                    width += dx
                }
                if (dy < 0 && viewPos.y == 0) {
                    height -= dy
                    curY -= dy
                } else if (dy >= 0 && maxY == panelSize.height) {
                    height += dy
                }
                scrollPane.horizontalScrollBar.value = viewPos.x + dx
                scrollPane.verticalScrollBar.value = viewPos.y + dy
            }

            Action.NOTHING -> {}
        }
        repaint()
    }

    override fun mouseMoved(e: MouseEvent) {
    }

    override fun mouseClicked(e: MouseEvent) {
        val centerX = getWidth().toDouble() / 2
        val centerY = getHeight().toDouble() / 2
        if (e.button == MouseEvent.BUTTON1) bSpline.addAnchorPoint(
            Point2D(
                (e.x - centerX) / indent, (centerY - e.y) / indent
            )
        )
        if (e.button == MouseEvent.BUTTON3) bSpline.deleteAnchorPoint()
        bSpline.createBSpline()
        repaint()
    }

    override fun mousePressed(e: MouseEvent) {
        curX = e.x
        curY = e.y
        checkPoint()
    }

    private fun checkPoint() {
        val anchorPoints = bSpline.getAnchorPoints()
        val centerX = getWidth() / 2
        val centerY = getHeight() / 2
        var i = 0
        for (p in anchorPoints) {
            val globalX = p.x * indent + centerX
            val globalY = centerY - p.y * indent
            if (curX <= globalX + Settings.RADIUS && curX >= globalX - Settings.RADIUS && curY <= globalY + Settings.RADIUS && curY >= globalY - Settings.RADIUS) {
                activePointIndex = i
                curAction = Action.POINT_MOVING
                break
            }
            i++
        }
    }

    override fun mouseReleased(e: MouseEvent) {
        curAction = Action.NOTHING
        repaint()
    }

    override fun mouseEntered(e: MouseEvent) {
    }

    override fun mouseExited(e: MouseEvent) {
    }

    override fun mouseWheelMoved(e: MouseWheelEvent) {
        val modifier = e.wheelRotation
        indent -= modifier
        repaint()
    }

    internal enum class Action {
        POINT_MOVING,
        CANVAS_MOVING,
        NOTHING
    }
}